File: tests/Blokemon.Game.Tests/CommandBoundaryTests.cs
using Blokemon.Game;
using Shouldly;

namespace Blokemon.Game.Tests;

public sealed class CommandBoundaryTests
{
    [Test]
    public async Task ReusingAnAcceptedCommand_IsRejectedWithoutChangingState()
    {
        var engine = MatchScenario.Engine();
        var state = MatchScenario.Started(engine.Start(MatchScenario.StartRequest()));
        var command = engine
            .GetLegalActions(state, MatchScenario.FirstPlayer)
            .First(action => action.Kind == LegalActionKind.ChooseOpening)
            .Command;
        var accepted = (CommandOutcome.Applied)engine.Apply(state, command);

        var duplicate = (CommandOutcome.Rejected)engine.Apply(accepted.State, command);

        duplicate.Rejection.Code.ShouldBe(CommandRejectionCode.DuplicateCommand);
        ReferenceEquals(duplicate.State, accepted.State).ShouldBeTrue();
        duplicate.State.ShouldBe(accepted.State);
    }

    [Test]
    public async Task ReusingACommandIdAtTheCurrentRevision_IsRejectedWithoutChangingState()
    {
        var engine = MatchScenario.Engine();
        var state = MatchScenario.Started(engine.Start(MatchScenario.StartRequest()));
        var command = engine
            .GetLegalActions(state, MatchScenario.FirstPlayer)
            .First(action => action.Kind == LegalActionKind.ChooseOpening)
            .Command;
        var accepted = (CommandOutcome.Applied)engine.Apply(state, command);
        var repeated = command with { ExpectedRevision = accepted.State.Revision };

        var duplicate = (CommandOutcome.Rejected)engine.Apply(accepted.State, repeated);

        duplicate.Rejection.Code.ShouldBe(CommandRejectionCode.DuplicateCommand);
        duplicate.State.ShouldBe(accepted.State);
    }

    [Test]
    public async Task StaleUniqueCommand_IsRejectedWithoutChangingState()
    {
        var engine = MatchScenario.Engine();
        var state = MatchScenario.Started(engine.Start(MatchScenario.StartRequest()));
        var command = engine
            .GetLegalActions(state, MatchScenario.FirstPlayer)
            .First(action => action.Kind == LegalActionKind.ChooseOpening)
            .Command;
        var accepted = (CommandOutcome.Applied)engine.Apply(state, command);
        var stale = command with { Id = new CommandId("unique-stale-command") };

        var rejected = (CommandOutcome.Rejected)engine.Apply(accepted.State, stale);

        rejected.Rejection.Code.ShouldBe(CommandRejectionCode.StaleRevision);
        rejected.State.ShouldBe(accepted.State);
    }
}

File: tests/Blokemon.Game.Tests/CommandBoundaryTests.fs
namespace Blokemon.Game.Tests

open Blokemon.Game
open FsUnit
open TUnit.Core

[<AutoOpen>]
module private CommandBoundaryFixtures =

    let openingCommand () =
        let engine = MatchScenario.Engine()
        let state = MatchScenario.Started(engine.Start(MatchScenario.StartRequest()))

        let command =
            engine.GetLegalActions(state, MatchScenario.FirstPlayer)
            |> Seq.find (fun action -> action.Kind = LegalActionKind.ChooseOpening)
            |> fun action -> action.Command

        engine, state, command

type CommandBoundaryTests() =

    [<Test>]
    member _.``replaying an accepted command should be refused without changing the state``() =
        let engine, state, command = openingCommand ()
        let accepted = MatchScenario.Applied(engine.Apply(state, command))

        let rejectedState, rejection =
            MatchScenario.Rejected(engine.Apply(accepted, command))

        rejection.Code |> should equal CommandRejectionCode.DuplicateCommand
        obj.ReferenceEquals(rejectedState, accepted) |> should be True
        rejectedState |> should equal accepted

    [<Test>]
    member _.``reusing a command identity at the new revision should be refused without changing the state``
        ()
        =
        let engine, state, command = openingCommand ()
        let accepted = MatchScenario.Applied(engine.Apply(state, command))

        let repeated =
            { command with
                ExpectedRevision = accepted.Revision }

        let rejectedState, rejection =
            MatchScenario.Rejected(engine.Apply(accepted, repeated))

        rejection.Code |> should equal CommandRejectionCode.DuplicateCommand
        rejectedState |> should equal accepted

    [<Test>]
    member _.``a fresh command against a stale revision should be refused without changing the state``
        ()
        =
        let engine, state, command = openingCommand ()
        let accepted = MatchScenario.Applied(engine.Apply(state, command))

        let stale =
            { command with
                Id = CommandId "unique-stale-command" }

        let rejectedState, rejection = MatchScenario.Rejected(engine.Apply(accepted, stale))

        rejection.Code |> should equal CommandRejectionCode.StaleRevision
        rejectedState |> should equal accepted
